p2p/security/tls/transport_test.go

```diff
@@ -175,6 +176,84 @@ func TestHandshakeSucceeds(t *testing.T) {
 	})
 }
 
+type testcase struct {
+	clientProtos   []protocol.ID
+	serverProtos   []protocol.ID
+	expectedResult string
+}
+
+func TestHandshakeWithNextProtoSucceeds(t *testing.T) {
+
+	tests := []testcase{
+		{clientProtos: nil, serverProtos: nil, expectedResult: ""},
+		{[]protocol.ID{"muxer1/1.0.0", "muxer2/1.0.1"}, []protocol.ID{"muxer2/1.0.1", "muxer1/1.0.0"}, "muxer2/1.0.1"},
+		{[]protocol.ID{"muxer1/1.0.0", "muxer2/1.0.1", "libp2p"}, []protocol.ID{"muxer2/1.0.1", "muxer1/1.0.0", "libp2p"}, "muxer2/1.0.1"},
+		{[]protocol.ID{"muxer1/1.0.0", "libp2p"}, []protocol.ID{"libp2p"}, ""},
+		{[]protocol.ID{"libp2p"}, []protocol.ID{"libp2p"}, ""},
+		{[]protocol.ID{"muxer1"}, []protocol.ID{}, ""},
+		{[]protocol.ID{}, []protocol.ID{"muxer1"}, ""},
+		{[]protocol.ID{"muxer2"}, []protocol.ID{"muxer1"}, ""},
+	}
+	numMuxers := len(tests)
+	var expectedMuxer string
+
+	clientID, clientKey := createPeer(t)
+	serverID, serverKey := createPeer(t)
+
+	handshake := func(t *testing.T, clientTransport *Transport, serverTransport *Transport) {
+		clientInsecureConn, serverInsecureConn := connect(t)
+
+		serverConnChan := make(chan sec.SecureConn)
+		go func() {
+			serverConn, err := serverTransport.SecureInbound(context.Background(), serverInsecureConn, "")
+			require.NoError(t, err)
+			serverConnChan <- serverConn
+		}()
+
+		clientConn, err := clientTransport.SecureOutbound(context.Background(), clientInsecureConn, serverID)
+		require.NoError(t, err)
+		defer clientConn.Close()
+
+		var serverConn sec.SecureConn
+		select {
+		case serverConn = <-serverConnChan:
+		case <-time.After(250 * time.Millisecond):
+			t.Fatal("expected the server to accept a connection")
+		}
+		defer serverConn.Close()
+
+		require.Equal(t, clientConn.LocalPeer(), clientID)
+		require.Equal(t, serverConn.LocalPeer(), serverID)
+		require.True(t, clientConn.LocalPrivateKey().Equals(clientKey), "client private key mismatch")
+		require.True(t, serverConn.LocalPrivateKey().Equals(serverKey), "server private key mismatch")
+		require.Equal(t, clientConn.RemotePeer(), serverID)
+		require.Equal(t, serverConn.RemotePeer(), clientID)
+		require.True(t, clientConn.RemotePublicKey().Equals(serverKey.GetPublic()), "server public key mismatch")
+		require.True(t, serverConn.RemotePublicKey().Equals(clientKey.GetPublic()), "client public key mismatch")
+		require.Equal(t, clientConn.ConnState().NextProto, expectedMuxer)
+		// exchange some data
+		_, err = serverConn.Write([]byte("foobar"))
+		require.NoError(t, err)
+		b := make([]byte, 6)
+		_, err = clientConn.Read(b)
+		require.NoError(t, err)
+		require.Equal(t, string(b), "foobar")
+	}
+
+	// Iterate through the NextProto combinations.
+	for i := 0; i < numMuxers; i++ {
```

Comment:
Nit: It's better to use a [for range](https://gobyexample.com/range) loop. This will also allow you to get rid of `numMuxers`.

Reply:
done.